### PR TITLE
zcash: init at 1.0.0

### PR DESCRIPTION
**Pull request by jb55**
###### Motivation for this change

zcash releases today, let's add it to nixos!
###### Things done
- [x] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
  - [x] NixOS
  - [ ] OS X
  - [x] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/top-level/all-packages.nix [hunk 1]
```diff
@@ -12068,8 +12070,17 @@ in
 
   altcoins = recurseIntoAttrs ( callPackage ../applications/altcoins {
     callPackage = newScope { boost = boost155; };
-  } );
+  } ) // (
+    let callPackage = newScope { boost = boost162; openssl = openssl_1_1_0; };
```
Comment (copumpkin) [inline pkgs/top-level/all-packages.nix]: If it really wants particular versions of boost and openssl, wouldn't it be easier to just use those parameter names in the expression?

File: pkgs/applications/altcoins/zcash/default.nix [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+, zlib, gtest, gmock, miniupnpc, callPackage, gmp, qt4, utillinux, protobuf, qrencode, libevent
+, withGui }:
+
+let snark = callPackage ./snark { };
+in
+with stdenv.lib;
+stdenv.mkDerivation rec{
+
+  name = "zcash" + (toString (optional (!withGui) "d")) + "-" + version;
+  version = "v1.0.0-rc4";
```
Comment (joachifm) [inline pkgs/applications/altcoins/zcash/default.nix]: Please remove the `v` from the version string.

Comment (sjmackenzie) [inline pkgs/applications/altcoins/zcash/default.nix]: did you try compile it without the `v`?

**Comment by jb55** [comment]
I wanted to ensure nested callPackage calls had the same versions of those
libs. I was having linking conflicts when linking libsnark without this. I
can try again perhaps I was mistaken.

**Comment by jb55** [comment]
v1.0.0 came out, will bump in a sec after some testing

**Comment by jb55** [comment]
There's also a [standalone stratum-miner](https://github.com/str4d/zcash/commits/standalone-miner) fork which I'll add as well since that's what most people probably want

File: pkgs/applications/altcoins/zcash/default.nix (hunk 2 above)
Comment (sjmackenzie) [inline pkgs/applications/altcoins/zcash/default.nix]: interesting, because I removed it and it didn't build. How very strange.

**Comment by sjmackenzie** [comment]
Yes please @jb55 I'd like the stratum-miner please!

**Comment by jb55** [comment]
hmm for some reason the stratum miner fork is crashing on mainnet...

**Comment by sjmackenzie** [comment]
What's in your `~/.zcash/zcash.conf`?

**Comment by sjmackenzie** [comment]
perchance are you getting errors like this:
```
...
2016-10-28 17:47:42 ZcashMiner started
2016-10-28 17:47:42 ZcashMiner started
2016-10-28 17:47:42 ZcashMiner started
2016-10-28 17:47:53 Loading addresses from DNS seeds (could take a while)
2016-10-28 17:48:19 26 addresses found from DNS seeds
2016-10-28 17:48:19 dnsseed thread exit
2016-10-28 17:48:19 connect() to 82.76.162.247:8233 failed after select(): Connection refused (111)
2016-10-28 17:48:42 connect() to 107.170.207.39:8233 failed after select(): Connection refused (111)
2016-10-28 17:52:39 connect() to 82.76.162.247:8233 failed after select(): Connection refused (111)
2016-10-28 17:59:14 socket recv error Connection reset by peer (104)
2016-10-28 17:59:27 connect() to 107.170.207.39:8233 failed after select(): Connection refused (111)
2016-10-28 18:04:03 connect() to 82.76.162.247:8233 failed after select(): Connection refused (111)
```

**Comment by jb55** [comment]
so it looks like crashing was just an issue with the pool I was using (suprnova) zmine.io works though

**Comment by sjmackenzie** [comment]
Okay great, might be a better idea to just default to the stratum-miner, hashrates are starting to soar, solo mining is already impossible.

**Comment by jb55** [comment]
done

**Comment by sjmackenzie** [comment]
whoohoo

**Comment by jb55** [comment]
http://zcash.flypool.org/ is another pool that seems to be working atm

**Comment by sjmackenzie** [comment]
Okay it built and executed, this is the contents of my `~/.zcash/zcash.conf` 
```
$ cat zcash.conf 
addnode=cn1-zcash.flypool.org:3333
rpcuser=zcHtaXQetFbFrwjmVmLhiYmCW9QPbQWYzcaWpq1h33BDAVmjNGmot2FFbVVXGqjRxKh6HNpg2gWwWBSwjLsYA7PFKCxb2Wg
rpcpassword=x
genproclimit=-1
gen=1
```
and I'm still getting 

```
You are running 8 mining threads.

Since starting this node 3 minutes, 51 seconds ago:
- You have validated 0 transactions!
- You have contributed 0.0000 Sol/s on average to the network solution rate.
- You have completed 0 Equihash solver runs.
```

the `~/.zcash/debug.log` looks like this:

```
2016-10-28 18:43:16 ZcashMiner started
2016-10-28 18:43:16 ZcashMiner started
2016-10-28 18:43:16 ZcashMiner started
2016-10-28 18:43:16 ZcashMiner started
2016-10-28 18:43:18 connect() to 82.76.162.247:8233 failed after select(): Connection refused (111)
2016-10-28 18:43:23 connect() to 107.170.207.39:8233 failed after select(): Connection refused (111)
2016-10-28 18:43:27 Loading addresses from DNS seeds (could take a while)
2016-10-28 18:43:43 connect() to 178.151.183.157:8233 failed after select(): Connection refused (111)
2016-10-28 18:44:04 26 addresses found from DNS seeds
2016-10-28 18:44:04 dnsseed thread exit
2016-10-28 18:44:43 socket recv error Connection reset by peer (104)
2016-10-28 18:44:57 socket recv error Connection reset by peer (104)
```
Do you see any problems? Something isn't right...

**Comment by jb55** [comment]
try adding:
```
equihashsolver=tromp
mainnet=1
```
to your conf

**Comment by jb55** [comment]
flypool started crashing on me for some reason, had to switch back to zmine + taddr

**Comment by sjmackenzie** [comment]
Okay I'll switch too, added your params and also noted this:
```
$ zcash-cli --version
Zcash RPC client version v1.0.0-rc4-gd2a221d
```
what is your version?

**Comment by jb55** [comment]
zcash-cli --version                                                                   ~
Zcash RPC client version v1.0.0-ged03ab9

**Comment by sjmackenzie** [comment]
okay, I just did a `nix-build`, whereas I should have `nix-env  still not getting sols/s on `1.0.0-ged03ab9`

could you please copy paste your `~/.zcash/zcash.conf` please.

**Comment by jb55** [comment]
```
equihashsolver=tromp
mainnet=1
genproclimit=24
addnode=mainnet.z.cash
rpcuser=user
rpcpassword=pass
```
is my zcash.conf

**Comment by jb55** [comment]
I'm running: 
`zcash-miner -equihashsolver=tromp -showmetric=1 -stratum=stratum+tcp://zmine.io:1337 -user=t1addr -genproclimit=24`

**Comment by jb55** [comment]
should look like this
```
Thank you for running a Zcash miner!
You're helping to strengthen the network :)

You are running 24 mining threads.

Since starting this node 12 minutes, 50 seconds ago:
- You have validated 0 transactions...
- You have contributed 12.6455 Sol/s on average to the network solution rate.
- You have completed 5124 Equihash solver runs.
- You have submitted 38 solutions!
  Accepted: 34, Rejected: 4, Failed: 0
```

**Comment by sjmackenzie** [comment]
Oh very interesting, I don't have `zcash-miner` only `zcashd` 
`<tabbing>`
```
[stewart@rivergod:~/.zcash]$ zcash
zcash-cli           zcashd              zcash-fetch-params  zcash-gtest         zcash-tx
```

**Comment by jb55** [comment]
is your nixpkgs on this commit? https://github.com/NixOS/nixpkgs/pull/19944/commits/44cdfbd52c9046e01de9f19166fff4d8f85eb462
also make sure to uninstall the old version...

**Comment by sjmackenzie** [comment]
no, it's `210b3b3`, but it's that for a reason :-), okay I see `zcash-miner` in the result folder. I'll use that!

**Comment by sjmackenzie** [comment]
I copied and pasted the changes manually.

**Comment by sjmackenzie** [comment]
okay, it's hashing nicely... btw you have a vicious machine with 24 threads. Thanks so much for this!

File: pkgs/top-level/all-packages.nix (hunk 1 above)
Comment (jb55) [inline pkgs/top-level/all-packages.nix]: sounds good to me

Comment (jb55) [inline pkgs/top-level/all-packages.nix]: let me know if [this looks good now](https://github.com/NixOS/nixpkgs/pull/19944/commits/e05082861a05700ef98eb2486d4714dea54d7211)

File: pkgs/top-level/all-packages.nix [hunk 3]
```diff
@@ -6715,6 +6716,7 @@ in
   db53 = callPackage ../development/libraries/db/db-5.3.nix { };
   db6 = db60;
   db60 = callPackage ../development/libraries/db/db-6.0.nix { };
+  db62 = callPackage ../development/libraries/db/db-6.2.nix { };
```
Comment (domenkozar) [inline pkgs/top-level/all-packages.nix]: Any reason not to bump 6.0 to 6.2?

**Comment by jb55** [comment]
I'm not planning on maintaining this, so closing for now.

**Comment by sjmackenzie** [comment]
@jb55 thanks for making this! Too bad maintainers couldn't grab the moment for this software while the iron was hot!

**Pull request closed without merging**